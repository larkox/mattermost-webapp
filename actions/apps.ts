// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {Client4} from 'mattermost-redux/client';
import {submitInteractiveDialog as submitInteractiveDialogRedux} from 'mattermost-redux/actions/integrations';
import {DialogSubmission} from 'mattermost-redux/types/integrations';
import {getCurrentChannelId} from 'mattermost-redux/selectors/entities/common';
import {ActionFunc, DispatchFunc, GetStateFunc} from 'mattermost-redux/types/actions';
import {AppCallResponseTypes, AppCallResponse, AppCall} from 'mattermost-redux/types/apps';

import {sendEphemeralPost} from 'actions/global_actions';
import {openInteractiveDialog} from 'plugins/interactive_dialog';
import {executeCommand} from 'actions/command';

export function doAppCall(call: AppCall): ActionFunc {
    return async (dispatch: DispatchFunc) => {
        const res = await Client4.executeAppCall(call) as AppCallResponse<any>; // TODO use unknown and check types (may need type guards)

        switch (res.type) {
        case AppCallResponseTypes.OK:
            if (res.markdown) {
                if (!call.context.channel_id) {
                    return {data: res};
                }

                sendEphemeralPost(res.markdown, call.context.channel_id, call.context.root_id);
            }
            return {data: res};
        case AppCallResponseTypes.MODAL:
            if (!res.data) {
                return {data: res};
            }
            res.data.app_id = call.context.app_id;
            openInteractiveDialog(res.data);
            return {data: res};
        case AppCallResponseTypes.COMMAND:
            res.data.args.command = res.data.command;
            return dispatch(executeCommand(res.data.command, res.data.args));
        }

        return {data: res};
    };
}

export function submitInteractiveDialog(dialog: DialogSubmission): ActionFunc {
    return (dispatch: DispatchFunc, getState: GetStateFunc) => {
        if (!dialog.submission.app_id) {
            return dispatch(submitInteractiveDialogRedux(dialog));
        }

        // Should `from` be used here instead of getting the channelID this way?
        // This interactive dialog could have been triggered by a command run in the RHS for example.
        // The previous Call or CallResponse would contain the correct channelID
        const channelID = getCurrentChannelId(getState());
        return dispatch(doAppCall({
            url: dialog.url,
            values: dialog.submission,
            context: {
                app_id: dialog.submission.app_id,
                channel_id: channelID,
            },
        }));
    };
}
