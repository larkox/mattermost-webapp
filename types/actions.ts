// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {ActionResult, GenericAction} from 'mattermost-redux/types/actions';
import { BatchAction } from 'redux-batched-actions';
import { GlobalState } from './store';

export type SuccessResult = {data: any;}

export type ErrorResult = {error: any;}

export function isSuccess(result: ActionResult): result is SuccessResult {
    return 'data' in result && !isError(result);
}

export function isError(result: ActionResult): result is ErrorResult {
    return Boolean((result as ErrorResult).error);
}
