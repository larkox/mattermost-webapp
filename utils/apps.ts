// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {AppCallResponseTypes} from 'mattermost-redux/constants/apps';
import {AppBinding, AppCall, AppCallValues, AppExpand} from 'mattermost-redux/types/apps';
import {GlobalState} from 'mattermost-redux/types/store';

export function appsEnabled(state: GlobalState) {// eslint-disable-line @typescript-eslint/no-unused-vars
    // TODO uncomment when featur flag is in place
    //return getConfig(state)?.['FeatureFlagApps' as keyof Partial<ClientConfig>];
    return true;
}

export function fillBindingsInformation(binding?: AppBinding) {
    if (!binding) {
        return;
    }

    binding.bindings?.forEach((b) => {
        // Propagate id down if not defined
        if (!b.app_id) {
            b.app_id = binding.app_id;
        }

        // Compose location
        b.location = binding.location + '/' + b.location;

        // Propagate call down if not defined
        if (!b.call) {
            b.call = binding.call;
        }

        fillBindingsInformation(b);
    });

    // Trim branches without app_id
    if (!binding.app_id) {
        binding.bindings = binding.bindings?.filter((v) => v.app_id);
    }

    // Trim branches without calls
    if (!binding.call) {
        binding.bindings = binding.bindings?.filter((v) => v.call);
    }

    // Pull up app_id if needed
    if (binding.bindings?.length && !binding.app_id) {
        binding.app_id = binding.bindings[0].app_id;
    }

    // Pull up call if needed
    if (binding.bindings?.length && !binding.call) {
        binding.call = binding.bindings[0].call;
    }
}

export function createCallRequest(
    call: AppCall,
    userId: string,
    appId: string,
    location: string,
    expand: AppExpand = {},
    channelId = '',
    postId = '',
    type = '',
    rawCommand = '',
    values?: AppCallValues,
): AppCall {
    return {
        type,
        path: call.path,
        context: {
            ...call.context,
            acting_user_id: userId,
            app_id: appId,
            channel_id: channelId,
            location,
            post_id: postId,
            user_id: userId,
        },
        values: {
            ...call.values,
            ...values,
        },
        expand: {
            ...call.expand,
            ...expand,
        },
        raw_command: rawCommand,
    };
}

export const makeCallErrorResponse = (errMessage: string) => {
    return {
        type: AppCallResponseTypes.ERROR,
        error: errMessage,
    };
};
