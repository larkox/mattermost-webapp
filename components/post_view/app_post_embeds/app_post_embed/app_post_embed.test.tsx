// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';
import {shallow} from 'enzyme';

import {MessageAttachment as MessageAttachmentType} from 'mattermost-redux/types/message_attachments';
import {Post} from 'mattermost-redux/types/posts';

import {AppBinding, AppPostEmbed} from 'mattermost-redux/types/apps';

import AppPostEmbedComponent from './app_post_embed';

describe('components/post_view/app_post_embeds/app_post_embed', () => {
    const post = {
        id: 'post_id',
        channel_id: 'channel_id',
    } as Post;

    const embed = {
        app_id: 'app_id',
        bindings: [] as AppBinding[],
        text: 'some text',
        title: 'some title',
    } as AppPostEmbed;

    const baseProps = {
        post,
        embed,
        currentRelativeTeamUrl: 'dummy_team',
    };

    test('should match snapshot', () => {
        const wrapper = shallow(<AppPostEmbedComponent {...baseProps}/>);
        expect(wrapper).toMatchSnapshot();
    });

    test('should match snapshot when the attachment has an emoji in the title', () => {
        const props = {
            ...baseProps,
            attachment: {
                title: 'Do you like :pizza:?',
            } as MessageAttachmentType,
        };

        const wrapper = shallow(<AppPostEmbedComponent {...props}/>);

        expect(wrapper).toMatchSnapshot();
    });

    test('should match snapshot when the attachment hasn\'t any emojis in the title', () => {
        const props = {
            ...baseProps,
            attachment: {
                title: 'Don\'t you like emojis?',
            } as MessageAttachmentType,
        };

        const wrapper = shallow(<AppPostEmbedComponent {...props}/>);

        expect(wrapper).toMatchSnapshot();
    });

    test('should match snapshot when the attachment has a link in the title', () => {
        const props = {
            ...baseProps,
            attachment: {
                title: 'Do you like https://mattermost.com?',
            } as MessageAttachmentType,
        };

        const wrapper = shallow(<AppPostEmbedComponent {...props}/>);

        expect(wrapper).toMatchSnapshot();
    });
});
