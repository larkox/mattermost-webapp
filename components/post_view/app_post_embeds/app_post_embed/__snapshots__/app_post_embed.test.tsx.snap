// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/post_view/app_post_embeds/app_post_embed should match snapshot 1`] = `
<div
  className="attachment"
  onClick={[Function]}
>
  <div
    className="attachment__content"
  >
    <div
      className="clearfix attachment__container attachment__container--"
    >
      <h1
        className="attachment__title"
      >
        <Connect(Markdown)
          message="some title"
          options={
            Object {
              "autolinkedUrlSchemes": Array [],
              "mentionHighlight": false,
              "renderer": LinkOnlyRenderer {
                "options": Object {},
              },
            }
          }
        />
      </h1>
      <div>
        <div
          className="attachment__body attachment__body--no_thumb"
        >
          <Connect(ShowMore)
            isAttachmentText={true}
            text="some text"
          >
            <Connect(Markdown)
              message="some text"
            />
          </Connect(ShowMore)>
        </div>
        <div
          style={
            Object {
              "clear": "both",
            }
          }
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`components/post_view/app_post_embeds/app_post_embed should match snapshot when the attachment has a link in the title 1`] = `
<div
  className="attachment"
  onClick={[Function]}
>
  <div
    className="attachment__content"
  >
    <div
      className="clearfix attachment__container attachment__container--"
    >
      <h1
        className="attachment__title"
      >
        <Connect(Markdown)
          message="some title"
          options={
            Object {
              "autolinkedUrlSchemes": Array [],
              "mentionHighlight": false,
              "renderer": LinkOnlyRenderer {
                "options": Object {},
              },
            }
          }
        />
      </h1>
      <div>
        <div
          className="attachment__body attachment__body--no_thumb"
        >
          <Connect(ShowMore)
            isAttachmentText={true}
            text="some text"
          >
            <Connect(Markdown)
              message="some text"
            />
          </Connect(ShowMore)>
        </div>
        <div
          style={
            Object {
              "clear": "both",
            }
          }
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`components/post_view/app_post_embeds/app_post_embed should match snapshot when the attachment has an emoji in the title 1`] = `
<div
  className="attachment"
  onClick={[Function]}
>
  <div
    className="attachment__content"
  >
    <div
      className="clearfix attachment__container attachment__container--"
    >
      <h1
        className="attachment__title"
      >
        <Connect(Markdown)
          message="some title"
          options={
            Object {
              "autolinkedUrlSchemes": Array [],
              "mentionHighlight": false,
              "renderer": LinkOnlyRenderer {
                "options": Object {},
              },
            }
          }
        />
      </h1>
      <div>
        <div
          className="attachment__body attachment__body--no_thumb"
        >
          <Connect(ShowMore)
            isAttachmentText={true}
            text="some text"
          >
            <Connect(Markdown)
              message="some text"
            />
          </Connect(ShowMore)>
        </div>
        <div
          style={
            Object {
              "clear": "both",
            }
          }
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`components/post_view/app_post_embeds/app_post_embed should match snapshot when the attachment hasn't any emojis in the title 1`] = `
<div
  className="attachment"
  onClick={[Function]}
>
  <div
    className="attachment__content"
  >
    <div
      className="clearfix attachment__container attachment__container--"
    >
      <h1
        className="attachment__title"
      >
        <Connect(Markdown)
          message="some title"
          options={
            Object {
              "autolinkedUrlSchemes": Array [],
              "mentionHighlight": false,
              "renderer": LinkOnlyRenderer {
                "options": Object {},
              },
            }
          }
        />
      </h1>
      <div>
        <div
          className="attachment__body attachment__body--no_thumb"
        >
          <Connect(ShowMore)
            isAttachmentText={true}
            text="some text"
          >
            <Connect(Markdown)
              message="some text"
            />
          </Connect(ShowMore)>
        </div>
        <div
          style={
            Object {
              "clear": "both",
            }
          }
        />
      </div>
    </div>
  </div>
</div>
`;
